create team management section of information gathering page. (#91)
wireapp/wire-docs: src/how-to/install/installation.md

# How to install wire

We have a pipeline in  `wire-server-deploy` producing container images, static
binaries, ansible playbooks, debian package sources and everything required to
install Wire.

On your machine (we call this the "admin host"), you need to have `docker`
installed (or any other compatible container runtime really, even though
instructions may need to be modified). See [how to install
docker](https://docker.com) for instructions.

Create a fresh workspace to download the artifacts:

```
cd ...  # you pick a good location!
```

Copy over ssh keys, or generate a key pair exclusively for this installation:

```
mkdir -p ./dot_ssh
cp ~/.ssh/id_rsa ./dot_ssh/
```

vs.

```
ssh-keygen -t ed25519 -a 100 -f ./dot_ssh/id_ed25519
ssh-add ./dot_ssh/id_ed25519
````

> Ensure the nodes your deploying to, as well as the bastion host accept this ssh key for user `root`:

Author: no root. username.

Reviewer: please push an edit commit on top.

> Ensure the nodes your deploying to, as well as the bastion host accept this ssh key for user `root`:

Author: ```suggestion
Ensure the nodes your deploying to, as well as any bastion host between you and them accept this ssh key for the user you are performing the install as:
```

```
# make sure the server accepts your ssh key for user root
# TODO: replace with ansible oneliner making use of the inventory

> ssh-copy-id -i ./dot_ssh/id_ed25519.pub root@<server>

Author: ```suggestion
ssh-copy-id -i ./dot_ssh/id_ed25519.pub <username>@<server>
```

   installed during other ansible playbook runs.
 - `charts`
   The charts themselves, as tarballs. We don't use an external helm
   repository, every helm chart dependency is resolved already.
 - `containers-system`
   These are the container images needed to bootstrap kubernetes itself
   (currently using kubespray)
 - `containers-helm`
   These are the container images our charts (and charts we depend on) refer to.
   Also come as tarballs, and are seeded like the system containers.
 - *containers-other*
   These are other container images, not deployed inside k8s. Currently only
   contains restund.
 - `debs`
   This acts as a self-contained dump of all packages required to install
   kubespray, as well as all other packages that are installed by ansible
   playbooks on nodes that don't run kubernetes.
   There's an ansible playbook copying these assets to an "assethost", starting
   a little webserver there serving it, and configuring all nodes to use it as
   a package repo.
 - `values`
   Contains helm chart values and secrets. Needs to be tweaked to the
   environment.

Provide a `ansible/inventory/offline/hosts.ini` configured to your environment.

> Copy over `hosts.ini.example` from that same folder.

Author: ```suggestion
Copy over `hosts.ini.example`  to `hosts.ini`, and edit it, following the instructions in that file.
```

Reviewer: please push an edit commit on top.

> Also, take a look at `ansible/inventory/offline/group_vars/all//offline.yml` to

Author: ```suggestion
Also, take a look at `ansible/inventory/offline/group_vars/all/offline.yml` to
```

Reviewer: please push an edit commit on top.

install Wire.

On your machine (we call this the "admin host"), you need to have `docker`
installed (or any other compatible container runtime really, even though
instructions may need to be modified). See [how to install
docker](https://docker.com) for instructions.

Create a fresh workspace to download the artifacts:

```
cd ...  # you pick a good location!
```

Obtain the latest artifacts from wire-server-deploy. Once
https://github.com/wireapp/wire-server-deploy/pull/363 is finished, these will
be in the "Releases" tab. Until then, go to the PRs "Checks" tab, and look for
the URL of the "assets" tarball linked from CI.

Extract the above listed artifacts into your workspace:

```
wget https://s3-eu-west-1.amazonaws.com/public.wire.com/artifacts/wire-server-deploy-static-77a1abdce68f77d6dd2ac53192825c6ac46c0aa0.tgz
tar xvzf wire-server-deploy-static-77a1abdce68f77d6dd2ac53192825c6ac46c0aa0.tgz
```

You should now have a directory named `assets` containing the offline deploy artifacts.

Copy over ssh keys, or generate a key pair exclusively for this installation:

```
mkdir -p ./assets/dot_ssh
cp ~/.ssh/id_rsa ./assets/dot_ssh/
```

vs.

> ssh-add ./assets/dot_ssh/id_ed25519

Author: ```suggestion
#create a key pair exclusively for this installation
```
ssh-keygen -t ed25519 -a 100 -f ../dot_ssh/id_ed25519
mkdir -p ./assets/dot_ssh
ssh-add ./assets/dot_ssh/id_ed25519
```

install Wire.

On your machine (we call this the "admin host"), you need to have `docker`
installed (or any other compatible container runtime really, even though
instructions may need to be modified). See [how to install
docker](https://docker.com) for instructions.

Create a fresh workspace to download the artifacts:

```
cd ...  # you pick a good location!
```

Obtain the latest artifacts from wire-server-deploy. Once
https://github.com/wireapp/wire-server-deploy/pull/363 is finished, these will
be in the "Releases" tab. Until then, go to the PRs "Checks" tab, and look for
the URL of the "assets" tarball linked from CI.

Extract the above listed artifacts into your workspace:

```
wget https://s3-eu-west-1.amazonaws.com/public.wire.com/artifacts/wire-server-deploy-static-77a1abdce68f77d6dd2ac53192825c6ac46c0aa0.tgz
tar xvzf wire-server-deploy-static-77a1abdce68f77d6dd2ac53192825c6ac46c0aa0.tgz
```

You should now have a directory named `assets` containing the offline deploy artifacts.

Copy over ssh keys, or generate a key pair exclusively for this installation:

```
mkdir -p ./assets/dot_ssh
cp ~/.ssh/id_rsa ./assets/dot_ssh/
```

vs.

> ssh-keygen -t ed25519 -a 100 -f ./assets/dot_ssh/id_ed25519

Author: ```suggestion
```
ssh-keygen -t ed25519 -a 100 -f ./assets/dot_ssh/id_ed25519
```

install Wire.

On your machine (we call this the "admin host"), you need to have `docker`
installed (or any other compatible container runtime really, even though
instructions may need to be modified). See [how to install
docker](https://docker.com) for instructions.

Create a fresh workspace to download the artifacts:

```
cd ...  # you pick a good location!
```

Obtain the latest artifacts from wire-server-deploy. Once
https://github.com/wireapp/wire-server-deploy/pull/363 is finished, these will
be in the "Releases" tab. Until then, go to the PRs "Checks" tab, and look for
the URL of the "assets" tarball linked from CI.

Extract the above listed artifacts into your workspace:

```
wget https://s3-eu-west-1.amazonaws.com/public.wire.com/artifacts/wire-server-deploy-static-77a1abdce68f77d6dd2ac53192825c6ac46c0aa0.tgz
tar xvzf wire-server-deploy-static-77a1abdce68f77d6dd2ac53192825c6ac46c0aa0.tgz
```

You should now have a directory named `assets` containing the offline deploy artifacts.

Copy over ssh keys, or generate a key pair exclusively for this installation:

```
mkdir -p ./assets/dot_ssh
cp ~/.ssh/id_rsa ./assets/dot_ssh/
```

vs.

> ```

Author: ```suggestion
```
mkdir -p ./assets/dot_ssh
```

   These are the container images needed to bootstrap kubernetes itself
   (currently using kubespray)
 - `containers-helm`
   These are the container images our charts (and charts we depend on) refer to.
   Also come as tarballs, and are seeded like the system containers.
 - *containers-other*
   These are other container images, not deployed inside k8s. Currently only
   contains restund.
 - `debs`
   This acts as a self-contained dump of all packages required to install
   kubespray, as well as all other packages that are installed by ansible
   playbooks on nodes that don't run kubernetes.
   There's an ansible playbook copying these assets to an "assethost", starting
   a little webserver there serving it, and configuring all nodes to use it as
   a package repo.
 - `values`
   Contains helm chart values and secrets. Needs to be tweaked to the
   environment.

Provide a `ansible/inventory/offline/hosts.ini` configured to your environment.
Copy over `hosts.ini.example`  to `hosts.ini`, and edit it, following the instructions in that file.

Also, take a look at `ansible/inventory/offline/group_vars/all/offline.yml` to
see if it matches your expectations.

> Copy over binaries and debs, serves assets from the asset host, and configures

Author: ```suggestion
Copy over binaries and debs, serves assets from the asset host, and configure
```

Copy over `hosts.ini.example`  to `hosts.ini`, and edit it, following the instructions in that file.
Don't forget to set a value for `restund_zrest_secret`, in case that is part of your inventory.

Also, take a look at `ansible/inventory/offline/group_vars/all/offline.yml` to
see if it matches your expectations.

Copy over binaries and debs, serves assets from the asset host, and configure
other hosts to fetch debs from it:

```
dapi ansible/setup-offline-sources.yml
```

Run kubespray until docker is installed and runs:

```
dapi ansible/kubernetes.yml --tags bastion,bootstrap-os,preinstall,container-engine
```

Now; run the restund playbook until docker is installed:
```
dapi ansible/restund.yml --tags docker
```

With docker being installed on all nodes that need it, seed all container images:

```
dapi ansible/seed-offline-docker.yml
```

Run the rest of kubespray:

```
dapi ansible/kubernetes.yml --skip-tags bootstrap-os,preinstall,container-engine
```

Copy over the kubeconfig from one master node:
```
d ansible -i ./ansible/inventory/offline/hosts.ini "kube-master[0]" -m fetch -a  "src=/root/.kube/config dest=ansible/kubeconfig flat=yes"
```

> Ensure the cluster comes up healthy. Copy TODO kubecfg from a master node, and put it to TODO. The container also contains kubectl, so check the node status:

Author: ```suggestion
Ensure the cluster comes up healthy. The container also contains kubectl, so check the node status:
```

```
d kubectl get nodes -owide
```
They should all report ready.

Now, deploy all other services which don't run in kubernetes:

```

> dapi ansible/restund.yaml

Author: ```suggestion
dapi ansible/restund.yml
```

> dapi ansible/cassandra.yaml

Author: ```suggestion
dapi ansible/cassandra.yml
```

> dapi ansible/elasticsearch.yaml

Author: ```suggestion
dapi ansible/elasticsearch.yml
```

> dapi ansible/minio.yaml

Author: ```suggestion
dapi ansible/minio.yml
```

```
TODO: add commands to verify things are good

Afterwards, run the following playbook to create helm values
```

> dapi ansible/helm_external.yaml

Author: ```suggestion
dapi ansible/helm_external.yml
```

   environment.

Provide a `ansible/inventory/offline/hosts.ini` configured to your environment.
Copy over `hosts.ini.example`  to `hosts.ini`, and edit it, following the instructions in that file.

In case that is part of your inventory, don't forget to set a value for `restund_zrest_secret`:
```
tr -dc 'a-zA-Z0-9' < /dev/urandom | fold -w 42 | head -n 1
```

Also, take a look at `ansible/inventory/offline/group_vars/all/offline.yml` to
see if it matches your expectations.

Copy over binaries and debs, serves assets from the asset host, and configure
other hosts to fetch debs from it:

```
dapi ansible/setup-offline-sources.yml
```

Run kubespray until docker is installed and runs:

```
dapi ansible/kubernetes.yml --tags bastion,bootstrap-os,preinstall,container-engine
```

Now; run the restund playbook until docker is installed:
```
dapi ansible/restund.yml --tags docker
```

With docker being installed on all nodes that need it, seed all container images:

```
dapi ansible/seed-offline-docker.yml
```

Run the rest of kubespray:

```
dapi ansible/kubernetes.yml --skip-tags bootstrap-os,preinstall,container-engine
```

Copy over the kubeconfig from one master node:
```

> d ansible -i ./ansible/inventory/offline/hosts.ini "kube-master[0]" -m fetch -a  "src=/root/.kube/config dest=ansible/kubeconfig flat=yes"

Reviewer 2: issue: This fails with `UNREACHABLE`. Probably because `assets/dot_ssh` is not mounted into the container being aliased with `d`

Reviewer: Scrolling up to L54 and L62, the aliases are created while being inside the `asserts/` folder, and `dot_ssh` should be created inside there as well, so this should work.

# How to install wire

We have a pipeline in  `wire-server-deploy` producing container images, static
binaries, ansible playbooks, debian package sources and everything required to
install Wire.

On your machine (we call this the "admin host"), you need to have `docker`
installed (or any other compatible container runtime really, even though
instructions may need to be modified). See [how to install
docker](https://docker.com) for instructions.

Create a fresh workspace to download the artifacts:

```
cd ...  # you pick a good location!
```

Obtain the latest artifacts from wire-server-deploy. Once
https://github.com/wireapp/wire-server-deploy/pull/363 is finished, these will
be in the "Releases" tab. Until then, go to the PRs "Checks" tab, and look for
the URL of the "assets" tarball linked from CI.

Extract the above listed artifacts into your workspace:

```

> wget https://s3-eu-west-1.amazonaws.com/public.wire.com/artifacts/wire-server-deploy-static-4a8bd3e21a078b378dcc577861947c04a180f976.tgz

Author: ```suggestion
wget https://s3-eu-west-1.amazonaws.com/public.wire.com/artifacts/wire-server-deploy-static-<HASH>.tgz
```

> tar xvzf wire-server-deploy-static-77a1abdce68f77d6dd2ac53192825c6ac46c0aa0.tgz

Author: ```suggestion
tar xvzf wire-server-deploy-static-<HASH>.tgz
```

```

> 

Author: ```suggestion
Where the HASH above is the hash of your deployment artifact, given to you by Wire, or acquired by looking at the above build job.

```

Reviewer 3: We should make the docs part of CI then this isn't a problem

# How to install wire

We have a pipeline in  `wire-server-deploy` producing container images, static
binaries, ansible playbooks, debian package sources and everything required to
install Wire.

On your machine (we call this the "admin host"), you need to have `docker`
installed (or any other compatible container runtime really, even though
instructions may need to be modified). See [how to install
docker](https://docker.com) for instructions.

> 

Author: ```suggestion

On ubuntu 18.04, connected to the internet:

apt install docker.io

```

tar xvzf wire-server-deploy-static-<HASH>.tgz
```
Where the HASH above is the hash of your deployment artifact, given to you by Wire, or acquired by looking at the above build job.

You should now have a directory named `assets` containing the offline deploy artifacts.

Copy over ssh keys, or generate a key pair exclusively for this installation:

```
mkdir -p ./assets/dot_ssh
cp ~/.ssh/id_rsa ./assets/dot_ssh/
```

vs.

```
ssh-add ./assets/dot_ssh/id_ed25519
```

Ensure the nodes you are deploying to, as well as any bastion host between you and them accept this ssh key for the user you are performing the install as:

```
# make sure the server accepts your ssh key for user root
# TODO: replace with ansible oneliner making use of the inventory
ssh-copy-id -i ./assets/dot_ssh/id_ed25519.pub <username>@<server>
```

Now `cd ./assets`. All the following operations should happen from the `assets` directory.

There's also a docker image containing the tooling inside this repo.

If you don't intend to develop *on wire-server-deploy itself*, you should load
this and register an alias:
```

> docker load -i ./containers-other/pxllp20fzzbafc179v4yjqijx9zw0254-docker-image-wire-server-deploy.tar.gz

Author: ```suggestion
docker load -i ./containers-other/*-docker-image-wire-server-deploy.tar.gz
```

# How to install wire

We have a pipeline in  `wire-server-deploy` producing container images, static
binaries, ansible playbooks, debian package sources and everything required to
install Wire.

On your machine (we call this the "admin host"), you need to have `docker`
installed (or any other compatible container runtime really, even though
instructions may need to be modified). See [how to install
docker](https://docker.com) for instructions.

On ubuntu 18.04, connected to the internet:

apt install docker.io

> Create a fresh workspace to download the artifacts:

Author: ```suggestion
Ensure the user you are using for the install has permission to run docker, or add 'sudo' to the docker commands below.

Create a fresh workspace to download the artifacts:
```

```
Where the HASH above is the hash of your deployment artifact, given to you by Wire, or acquired by looking at the above build job.

You should now have a directory named `assets` containing the offline deploy artifacts.

Copy over ssh keys, or generate a key pair exclusively for this installation:

```
mkdir -p ./assets/dot_ssh
cp ~/.ssh/id_rsa ./assets/dot_ssh/
```

vs.

```
ssh-add ./assets/dot_ssh/id_ed25519
```

Ensure the nodes you are deploying to, as well as any bastion host between you and them accept this ssh key for the user you are performing the install as:

```
# make sure the server accepts your ssh key for user root
# TODO: replace with ansible oneliner making use of the inventory
ssh-copy-id -i ./assets/dot_ssh/id_ed25519.pub <username>@<server>
```

Now `cd ./assets`. All the following operations should happen from the `assets` directory.

There's also a docker image containing the tooling inside this repo.

If you don't intend to develop *on wire-server-deploy itself*, you should load
this and register an alias:
```
docker load -i ./containers-other/pxllp20fzzbafc179v4yjqijx9zw0254-docker-image-wire-server-deploy.tar.gz

> alias d="docker run -it --network=host -v $PWD:/wire-server-deploy quay.io/wire/wire-server-deploy:pxllp20fzzbafc179v4yjqijx9zw0254"

Author: ```suggestion
alias d="docker run -it --network=host -v $PWD:/wire-server-deploy quay.io/wire/wire-server-deploy:<container_ID>"
```

> alias dapi="d ansible-playbook -i ansible/inventory/offline/hosts.ini --private-key ./dot_ssh/id_ed25519"

Author: assumes ed25519

Reviewer 3: where? the playbook just shells out to ssh. you could use any other key here

Author: sure, but we STATE a key type, without telling the user a thing.

Reviewer 3: I still don't understand? The only thing we state is a path to an SSH key.  The file could contain anything; ansible doesn't care

Reviewer 3: Also I think it's totally fine to namedrop `ed25519` wherever we can =) people shouldn't be using RSA keys in 2020.

   repository, every helm chart dependency is resolved already.
 - `containers-system`
   These are the container images needed to bootstrap kubernetes itself
   (currently using kubespray)
 - `containers-helm`
   These are the container images our charts (and charts we depend on) refer to.
   Also come as tarballs, and are seeded like the system containers.
 - *containers-other*
   These are other container images, not deployed inside k8s. Currently, only
   contains restund.
 - `debs`
   This acts as a self-contained dump of all packages required to install
   kubespray, as well as all other packages that are installed by ansible
   playbooks on nodes that don't run kubernetes.
   There's an ansible playbook copying these assets to an "assethost", starting
   a little webserver there serving it, and configuring all nodes to use it as
   a package repo.
 - `values`
   Contains helm chart values and secrets. Needs to be tweaked to the
   environment.

Provide a `ansible/inventory/offline/hosts.ini` configured to your environment.
Copy over `hosts.ini.example`  to `hosts.ini`, and edit it, following the instructions in that file.

> In case that is part of your inventory, don't forget to set a value for `restund_zrest_secret`:

Author: ```suggestion
If you are installing restund (TURN) services, don't forget to set a value for `restund_zrest_secret`:
```

```
tr -dc 'a-zA-Z0-9' < /dev/urandom | fold -w 42 | head -n 1
```

Also, take a look at `ansible/inventory/offline/group_vars/all/offline.yml` to
see if it matches your expectations.

Copy over binaries and debs, serves assets from the asset host, and configure
other hosts to fetch debs from it:

```
dapi ansible/setup-offline-sources.yml
```

Run kubespray until docker is installed and runs:

```
dapi ansible/kubernetes.yml --tags bastion,bootstrap-os,preinstall,container-engine
```

Now; run the restund playbook until docker is installed:
```
dapi ansible/restund.yml --tags docker
```

With docker being installed on all nodes that need it, seed all container images:

```
dapi ansible/seed-offline-docker.yml
```

Run the rest of kubespray:

```
dapi ansible/kubernetes.yml --skip-tags bootstrap-os,preinstall,container-engine
```

Copy over the kubeconfig from one master node:
```

> d ansible -i ./ansible/inventory/offline/hosts.ini "kube-master[0]" -m fetch -a  "src=/root/.kube/config dest=ansible/kubeconfig flat=yes"

Author: ```suggestion
d ansible -i ./ansible/inventory/offline/hosts.ini "kube-master[0]" --private-key ./dot_ssh/id_rsa -m fetch -a  "src=/root/.kube/config dest=ansible/kubeconfig flat=yes"
```

```
d kubectl get nodes -owide
```
They should all report ready.

Now, deploy all other services which don't run in kubernetes:

```
dapi ansible/restund.yml
dapi ansible/cassandra.yml
dapi ansible/elasticsearch.yml
dapi ansible/minio.yml
```
TODO: add commands to verify things are good

Afterwards, run the following playbook to create helm values
```
dapi ansible/helm_external.yml
```

Write other values by copying over TODO

Use Helm to apply the various charts in order to install the Wire platform

TODO: update https://docs.wire.com/how-to/install/helm-prod.html to not add the iptables hack

```
d helm install cassandra-external ./charts/cassandra-external --values ./values/cassandra-external/values.yaml
d helm install elasticsearch-external ./charts/elasticsearch-external --values ./values/elasticsearch-external/values.yaml
d helm install minio-external ./charts/minio-external --values ./values/minio-external/values.yaml
d helm install fake-aws ./charts/fake-aws --values ./values/fake-aws/values.yaml
d helm install demo-smtp ./charts/demo-smtp --values ./values/demo-smtp/values.yaml

> d helm install redis-ephemeral ./charts/nginx-ingress-controller

Author: ```suggestion
d helm install redis-ephemeral ./charts/redis-ephemeral
```
